### ISSUE TITLE
Proper TS typing when using from React class component (and using mst-gql with class-based React comps)

### ISSUE THREAD

**Issue by fullofcaffeine** [mobxjs/mst-gql#132]
Hi!

I'm trying to figure out the usage patterns for class-based React components (most, if not all, the examples are based of React hooks) and I did manage to make it work (I could call a mutation), but I can't get TS to get the proper typed context.

I have this entrypoint:

```typescript
import { configureStore } from '../shared/store';
import App from '../shared/App';
import createHistory from '../shared/store/history';
import { initializeStore } from '../utils/initModels';

const history = createHistory();

import { RootStore, StoreContext, RootStoreType } from "../models";

// history MUST be passed here if you want syncing between server on initial route
const store: RootStoreType =
    window.store ||
    initializeStore(false, window.__PRELOADED_STATE__);

function Main() {
  return (
    <StoreContext.Provider value={store}>
      <Router history={history}>
          <HelmetProvider>
            <ThemeProvider theme={theme}>
              <App />
            </ThemeProvider>
          </HelmetProvider>
      </Router>
    </StoreContext.Provider>
  );
}

ReactDOM.hydrate(
  <Main/>,
  document.getElementById('app')
);
````

And the `App.tsx`  component:

```typescript
interface IMatchParams { name: string; }
interface IProps extends RouteComponentProps<IMatchParams> {};

class App extends Component<IProps, {}> {
  constructor(props) {
    super(props);
  }
  public render() {
    const currentPath = this.props.location.pathname;
    if (__SERVER__) console.log('foofofofo', currentPath);
    return (
      <>
        <Helmet
          defaultTitle="Foobar"
        />
        <Topbar currentPath={currentPath}/>
        <Switch>
          <Route exact path='/' component={Home} />
          <Route exact path='/signup' component={SignUp} />
          <Route exact path='/signin' component={SignIn} />
          <Route exact path='/mst' component={MstTest} />
          <Route component={Error404} />
        </Switch>
      </>
    );
  }
}

export default withRouter(App);
``` 
Then in the `SignUp`  component, I have this:

```typescript
@observer
class SignUp extends React.Component<IProps, {}> {
  static contextType = StoreContext;

  constructor(props) {
    super(props);
  }

  public render() {
    const { classes } = this.props;
    console.log(this.context);
    return (
      <Container component="main" maxWidth="xs">
        <CssBaseline />
        <div className={classes.paper}>
          <Avatar className={classes.avatar}>
            <LockOutlinedIcon />
          </Avatar>
          <Typography component="h1" variant="h5">
            Sign up
        </Typography>
          <form className={classes.form} noValidate onSubmit={
            event => {
              event.preventDefault();
             /*
              Cannot get completion below, since `context`  type is `any` :( 
             */
              this.context.mutateRegister({
                firstName: 'foo',
                lastName: 'bar',
                email: 'foon@asd.com3',
                password: 'fooo'
              });
            }
          }>
          /* (...) omitted */
    );
  }
}

export default withStyles(styles)(SignUp);
```

**For some reason, the type of `this.context` is == `any` so I don't get any typing info for it nor for the `mutateRegister` method.  Why? How could I fix this?**

Also, it'd be really nice if someone could (I'd do if I knew!) add an example for class React component for each of the Hook examples present in the README. It'd really help a lot for those of us who don't use Hooks :)

Thanks in advance!

**Comment by odjhey** [comment]
We might not be able to leverage `useQuery` on class components due to it being built on react hooks. https://reactjs.org/docs/hooks-faq.html#should-i-use-hooks-classes-or-a-mix-of-both
You can prolly solve this by create a 'wrapper' (HOC), written using the functional approach, then wrap your class based component before using it.

For the typing part, you might get the types by using a consumer of StoreContext
https://reactjs.org/docs/context.html#contextconsumer

```typescript
<StoreContext.Consumer>
  {( store ) /* This can be typed against your StoreType */ 
    => {
    return <>  /* Render your view here*/  </>
  }}
</StoreContext.Consumer>
```

**Comment by fullofcaffeine** [comment]
> We might not be able to leverage useQuery on class components due to it being built with react hooks. https://reactjs.org/docs/hooks-faq.html#should-i-use-hooks-classes-or-a-mix-of-both
> You can prolly solve this by create a 'wrapper' (HOC), written using the functional approach, then wrap your class based component before using it.

That's basically a way to expose the hook to a class right? That's surely interesting and a technique that might come in handy in the future. However, I shouldn't need to do that for `mst-gql` right? I mean, using the StoreContext should be enough to provide the same functionality, or am I wrong?

> For the typing part, you may get the types by using a consumer of StoreContext
> https://reactjs.org/docs/context.html#contextconsumer

Oh nice! I'll try it out. However, do you know why the typings didn't work in my case? I'd love to know what I did wrong, because while the typings didn't work, the actual code runs fine. Why do I need to wrap it in a consumer just to get the typings?

**Comment by fullofcaffeine** [comment]
@odjhey Found out the missing piece. There's a full-blown example here:

https://github.com/piotrwitek/react-redux-typescript-guide/issues/140

It's for Redux, but the context part is framework agnostic.

TL;DR: For TS to be able to type the context, this line is needed:

```typescript
  context!: React.ContextType<typeof StoreContext>;
```

I haven't tried your approach with the consumer JSX tag, so not sure if typings work there out-of-the-box, but that might come in handy later. Thanks a a lot for the help!
